File: sorting/merge/merge.c
/* C program for merge sort algorithm */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

/*	Used to merge two subarrays of given array
	first	is arr[l..m]
	second	is arr[m+1..r]
*/

void merge(int arr[], int l, int m, int r) {
	int i, j, k;
	int n1 = m - l + 1;
	int n2 = r - m;

	/* Creates two temporary arrays */
	int L[n1], R[n2];

	/* Copies first	half of the data to array L[] */
	for (i = 0; i < n1; i++)
		L[i] = arr[l + i];

	/* Copies second half of the data to array R[] */
	for (j = 0; j < n2; j++)
		R[j] = arr[m + 1 + j];

	i = 0; // Initial index of L
	j = 0; // Initital index of R
	k = l; // Initial index of merged array

	/* Merges the temporary arrays */
	while (i < n1 && j < n2) {
		if (L[i] <= R[j]) {
			arr[k] = L[i];
			i++;
		}else {
			arr[k] = R[j];
			j++;
		}
		k++;
	}

	/* Copies the remaining elements of L */
	while (i < n1) {
		arr[k] = L[i];
		i++;
		k++;
	}

	/* Copies the remaining elements of R */
	while (j < n2) {
		arr[k] = R[j];
		j++;
		k++;
	}
}

/*  L is the left index and R is the right index */
void mergeSort(int arr[], int l, int r) {
    if (l < r) {
        // Same as (l+r)/2, but avoids overflow for
        // large l and h
        int m = l + (r - l) / 2;

        // Sort first and second halves
        mergeSort(arr, l, m);
        mergeSort(arr, m + 1, r);

        merge(arr, l, m, r);
    } 
}

/* Spits out the contents of the array*/
void printArray(int A[], int size) {
	int i;
	for(i = 0; i < size; i++) {
		printf("%d\n", A[i]);
	}
	printf("\n");
}

int main() {

	int arr[12], i;
	int arr_size = sizeof(arr) / sizeof(int);

    /* Uses epoch time as a seed for rand() */
    srand(time(NULL));
    for (i=0; i<11; i++) {
    	arr[i] = rand() % 25;
    }

    printf("Given array is \n"); 
    printArray(arr, arr_size); 

    mergeSort(arr, 0, arr_size - 1); 

 	printf("\nSorted array is \n"); 
	printArray(arr, arr_size); 
    return 0; 
}

File: sorting/merge/merge.py
def merge_sort(arr):
	""" Simple merge sort algorithm. For every input array of length >= 1:
			mergesort performs at most:
				6n log2 n + 6
n	"""
	if len(arr) > 1:
		mid = len(arr) // 2
		L = arr[:mid]
		R = arr[mid:]

		merge_sort(L)
		merge_sort(R)

		i = j = k = 0

		while i < len(L) and j < len(R):
			if L[i] < R[j]:
				arr[k] = L[i]
				i += 1
			else:
				arr[k] = R[j]
				j += 1
			k += 1

		while i < len(L):
			arr[k] = L[i]
			i += 1
			k += 1

		while j < len(R):
			arr[k] = R[j]
			j += 1
			k += 1


array = [2, 20, 1, 10, 3, 40, 4]

#	Unsorted
for i in array:
	print(i)		

merge_sort(array)

for i in array:
	print(i)
